### ISSUE TITLE
Unexpected token when using standard lexer

### ISSUE THREAD

**Issue by LKajan** [erezsh/lark#90]
I'm trying to parse a file line by line. Each line contains multiple optional parts and few mandatory parts. Parts have often multiple valid formats. I got this working with dynamic lexer but it takes over half a second per row. Now I'm trying to use standard lexer but I'm facing problems even with a really simple test.  
This must be just something that I don't understand.

Here's a sample script:

```python
grammar = r'''
    work_row : [date] work_time

    day: /\d{1,2}/
    month: /\d{1,2}/
    year: /\d{2}|\d{4}/
    date: day "." month "." year?
    hours: /\d+/
    minutes: /\d+/
    work_time: hours "h" (minutes "min")?

    %import common.WS
    %ignore WS
'''

l = Lark(grammar, start='work_row', lexer='standard')

print( l.parse("27.1.18 2h30min").pretty() )
```

This gives an error:

```
lark.common.UnexpectedToken: Unexpected token Token(__DOT, '.') at line 1, column 2.
Expected: {'__H3'}
Context: <no context>
```

**Comment by erezsh** [comment]
I'm surprised that the dynamic lexer takes so long for this grammar. Or is it only part of a larger grammar?

The standard lexer is by definition deterministic and grammar agnostic, which means it can't tell when `/\d{1,2}/` is appropriate and when `/\d+/` is, etc. It just chooses arbitrarily.

You should try using LALR(1) with the contextual lexer:

    l = Lark(grammar, start='work_row', parser='lalr', lexer='contextual')

But I expect there will still be some collisions. You should consider turning `date` into a terminal: 

    DATE: DAY "." MONTH "." YEAR?

same with `work_time`. You will have to "parse" them again in the tree, but it's a small cost considering you will gain parse speed and avoid many possible collisions.

**Comment by LKajan** [comment]
Thank you for the help and sorry for the delayed response!

Yes this was just a part of a larger grammar.

I got the grammar to work by replacing everythink with terminals.  
As you said now I just have to parse each token separately.

I also managed to improve my original Earley grammar and it now completes in .04 - .07 seconds. Lalr still outperforms. Lalr makes the job in 0.001 seconds, Yay! (without parsing the tokens though.)

Thanks for your help!
